File: Irrgarten_Ruby/game.rb
# encoding: UTF-8

require_relative 'dice'
require_relative 'player'
require_relative 'monster'
require_relative 'labyrinth'
require_relative 'game_state'
require_relative 'game_character'
require_relative 'orientation'

# Este módulo principal contiene las clases y módulos que representan el juego **Irrgarten**.
#
# @author Joaquin Avilés de la Fuente
# @author Arturo Olivares Martos
module Irrgarten

    # Clase que representa el juego Irrgarten. Esta clase se encarga de gestionar el estado del juego,
    # las acciones de los jugadores y los monstruos, y de generar un estado del juego que pueda ser
    # consultado por la interfaz de usuario.
    #
    # @author Joaquin Avilés de la Fuente
    # @author Arturo Olivares Martos
    class Game
        @@MAX_ROUNDS = 10 # Número máximo de rondas de cada combate

        ####---------------------- PERSONALIZACIÓN LABERINTO ----------------

        # Número de filas del laberinto
        @@ROWS = 4

        # Número de columnas del laberinto
        @@COLS = 4

        # Información de los monstruos a añadir al laberinto
        @@MONSTER_INIT = [
            ["Monster 0", 0, 0],
            ["Monster 1", 1, 1],
            ["Monster 2", 2, 2],
            ["Monster 3", 6, 7]
        ]

        # Información de los bloques de obstáculos a añadir al laberinto
        @@BLOCKS = [
            [Orientation::HORIZONTAL,   7, 7,   3],
            [Orientation::VERTICAL,     5, 5,   4],
            [Orientation::HORIZONTAL,   3, 5,   8],
            [Orientation::VERTICAL,     6, 5,   3]
        ]
        #### ---------------------- PERSONALIZACIÓN LABERINTO ----------------

        # Constructor de la clase Game. Inicializa los atributos de la clase.
        #
        # @param n_players [int] número de jugadores
        def initialize(n_players)

            @players = Array.new
            @monsters = Array.new

            # Inicializa el array de jugadores
            n_players.times do |i|
                @players.push(Player.new(i.to_s, Dice.random_intelligence, Dice.random_strength))
            end

            @current_player_index = Dice.who_starts(n_players)  # Índice del jugador que comienza la partida
            @current_player = @players[@current_player_index]    # Jugador que tiene el turno actual

            # Inicializa el laberinto
            @labyrinth = Labyrinth.new(@@ROWS, @@COLS, Dice.random_pos(@@ROWS), Dice.random_pos(@@COLS))
            configure_labyrinth()
            @labyrinth.spread_players(@players)

            @log = "- Game just started.\n"

        end

        # Delega en el método del laberinto que indica si hay un ganador.
        #
        # @return [boolean] **true** si hay un ganador, **false** en caso contrario
        def finished
            return @labyrinth.have_a_winner
        end

        # Método que hace una ronda de un jugador, en la que el jugador avanza una
        # casilla (si está vivo, sino se intenta resucitar) y se combate si está en
        # la misma casilla que un monstruo. Finalmente se comprueba si ha finalizado el
        # juego.
        #
        # @param preferred_direction [Directions] dirección hacia la que se
        # quiere desplazar el jugador
        #
        # @return [boolean] devuelve true si finalizo el juego, false en caso contrario
        def next_step(preferred_direction)
            @log = ""

            if(!@current_player.dead)
                direction = self.actual_direction(preferred_direction)

                # Se comprueba si la dirección a moverse ha sido la querida
                if(direction != preferred_direction)
                    self.log_player_no_orders
                end

                monster = @labyrinth.put_player(direction, @current_player)
                if(monster == nil)
                    self.log_no_monster
                else
                    winner = self.combat(monster)
                    self.manage_reward(winner)
                end
            else
                self.manage_resurrection
            end

            # Se comprueba si ha ganado ya alguien
            end_game = self.finished()
            if(!end_game)
                self.next_player
            end

            return end_game
        end

        # Genera una instancia de GameState integrando toda la información del estado del juego.
        #
        # @return [GameState] Objeto GameState con la información del estado del juego
        def game_state
            info_players = ""
            @players.each do |player|
                info_players += player.to_s + "\n"
            end

            info_monsters = ""
            @monsters.each do |monster|
                info_monsters += monster.to_s + "\n"
            end

            return GameState.new(@labyrinth.to_s, info_players, info_monsters, @current_player_index, self.finished, @log)
        end

        private

        # Configura el laberinto añadiendo bloques de obstáculos y monstruos.
        # Los monstruos, además de en el laberinto son guardados en el contenedor propio de esta clase para
        # este tipo de objetos.
        def configure_labyrinth

            # Añade los bloques de obstáculos al laberinto
            @@BLOCKS.each do |block|
                orientation, row, col, length = block
                @labyrinth.add_block(orientation, row, col, length)
            end


            # Añade los monstruos al laberinto y al contenedor de monstruos
            @@MONSTER_INIT.each do |monster_info|
                name, row, col = monster_info

                monster = Monster.new(name, Dice.random_intelligence, Dice.random_strength)

                @monsters.push(monster)
                @labyrinth.add_monster(row, col, monster)
            end
        end

        # Actualiza los dos atributos que indican el jugador con el turno pasando al siguiente jugador.
        def next_player
            @current_player_index = (@current_player_index + 1) % @players.size
            @current_player = @players[@current_player_index]
        end

        # Comprueba si la dirección preferred_direction es válida para el actual jugador,
        # delegando para ello en el método move de Player. Devuelve la dirección hacia la que se
        # desplazará el jugador actual (puede que no sea la preferida)
        # @see Player#move
        # @param preferred_direction [Directions] dirección hacia la que se pretende desplazar el
        # jugador actual
        #
        # @return [Directions] dirección a la que se desplazará el jugador actual
        # (puede que no sea la preferida)
        def actual_direction(preferred_direction)
            current_row = @current_player.row
            currect_col = @current_player.col
            valid_moves = @labyrinth.valid_moves(current_row, currect_col)

            output = @current_player.move(preferred_direction, valid_moves)

            return output
        end

        # Método que desarrolla un combate entre el jugador actual y un monstruo, donde es el
        # jugador el primero que empieza atacando. Además, delega en el método defend y attack
        # tanto de monster como de player. Finalmente escribe un mensaje con el número de rondas
        # que se han dado y devuelve el ganador.
        # @see Player#defend
        # @see Player#attack
        # @see Monster#defend
        # @see Monster#attack
        #
        # @param monster [Monster] monstruo con el que combate el jugador actual
        #
        # @return [GameCharacter] devuelve el ganador (PLAYER o MONSTER)
        def combat(monster)
            # Inicializamos los valores
            rounds = 0
            winner = GameCharacter::PLAYER

            # Comienza el jugador atacando
            lose = monster.defend(@current_player.attack)
            # Bucle hasta que finalice el número de rondas posible o haya perdido
            # el monstruo
            while ( (!lose) && (rounds<@@MAX_ROUNDS) )
                winner = GameCharacter::MONSTER
                rounds += 1

                # Turno del monstruo de atacar al jugador
                lose = @current_player.defend(monster.attack())
                if(!lose)
                    winner = GameCharacter::PLAYER
                    lose = monster.defend(@current_player.attack())
                end
            end

            self.log_rounds(rounds, @@MAX_ROUNDS)
            # Devolvemos el ganador
            return winner
        end

        # Comprueba quien fue el ganador del combate. Si fue el jugador se le otorga
        # una recompensa con el metodo receive_reward de Player, y en función de quien gane
        # se muestra un mensaje u otro.
        # @see Player#receive_reward
        #
        # @param winner [GameCharacter] ganador del combate
        def manage_reward(winner)
            if (winner == GameCharacter::PLAYER)
                @current_player.receive_reward
                log_player_won
            else
                log_monster_won
            end
        end

        # Comprueba mediante el medoto resurrect_player de Dice si el jugador actual revivirá.
        # En caso de que lo haga, se muestra un mensaje, sino se pasa su turno y se muestra el mensaje
        # correspondiente
        def manage_resurrection()

            if(Dice.resurrect_player())
                @current_player.resurrect
                log_resurrected
            else
                log_player_skip_turn

            end
        end

        # Añade al final del atributo log el mensaje indicando que el jugador ha ganado el combate.
        def log_player_won
            @log += "- Player #{@current_player_index} won the fight.\n"
        end

        # Añade al final del atributo log el mensaje indicando que el monstruo ha ganado el combate.
        def log_monster_won
            @log += "- Monster won the fight.\n"
        end

        # Añade al final del atributo log el mensaje indicando que el jugador ha resucitado.
        def log_resurrected
            @log += "- Player #{@current_player_index} resurrected.\n"
        end

        # Añade al final del atributo log el mensaje indicando que el jugador ha perdido el turno por estar muerto.
        def log_player_skip_turn
            @log += "- Player #{@current_player_index} skipped turn (is dead).\n"
        end

        # Añade al final del atributo log el mensaje indicando que el jugador no ha
        # seguido las instrucciones del jugador humano (no fue posible).
        def log_player_no_orders
            @log += "- Player #{@current_player_index} didn't follow orders, it was not possible.\n"
        end

        # Añade al final del atributo log el mensaje indicando que el jugador se ha movido a una casilla vacía o no le ha sido posible moverse.
        def log_no_monster
            @log += "- Player #{@current_player_index} moved to an empty cell or it was not possible to move.\n"
        end

        # Añade al final del atributo log el mensaje indicando que se han producido rounds de max rondas de combate.
        #
        # @param rounds [int] número de rondas de combate que ya se han producido
        # @param max [int] número máximo de rondas de combate
        def log_rounds(rounds, max)
            @log += "- Rounds: #{rounds}/#{max}.\n"
        end

    end
end

File: Irrgarten_Java/src/irrgarten/Game.java
package irrgarten;

import java.util.ArrayList;

/**
 * Clase que representa el juego Irrgarten. Esta clase se encarga de gestionar el estado del juego,
 * las acciones de los jugadores y los monstruos, y de generar un estado del juego que pueda ser
 * consultado por la interfaz de usuario.
 * 
 * @author Joaquín Avilés de la Fuente
 * @author Arturo Olivares Martos
 */
public class Game {
    /**
     * Número máximo de rondas que puede durar un combate.
     */
    private static final int MAX_ROUNDS=10;
    
// ---------------------- PERSONALIZACIÓN LABERINTO ---------------- //
    // Tamaño del laberinto
    /**
     * Número de filas del laberinto.
     */
    private static final int ROWS=12;
    /**
     * Número de columnas del laberinto.
     */
    private static final int COLS=12;
    

    // Monstruos a incluir
    /**
     * Número de monstruos a incluir en el laberinto.
     */
    private static final int NUM_MONSTERS=5;
    /**
     * Posiciones iniciales de los monstruos en el laberinto.
     */
    private static final int [][] INIT_MONSTERS= {{0,0}, {1,1}, {2,2}, {3,3}, {0,3}};
    
    // Bloques a incluir
    /**
     * Número de bloques a incluir en el laberinto.
     */
    private static final int NUM_BLOCKS=4;
    // Object es la clase raíz de la jerarquía de Java, por lo que todas las clases
    // existentes se pueden definir como Object, haciendo después el
    // correspondiente casting para usar el tipo de dato que queramos
    // y sea el correcto con el que hay
    /**
     * Bloques iniciales a incluir en el laberinto.
     */
    private static final Object [][] INIT_BLOCKS=
    { {Orientation.HORIZONTAL,0,0,3}, {Orientation.HORIZONTAL,3,3,5}, 
      {Orientation.VERTICAL,  5,6,2}, {Orientation.VERTICAL,  6,5,3} };

// ---------------------- PERSONALIZACIÓN LABERINTO ---------------- //
    
    /**
     * Índice del jugador actual.
     */
    private int currentPlayerIndex;
    /**
     * Jugador actual.
     */
    private Player currentPlayer;

    /**
     * Log del juego. Informa de las acciones que se han producido en el juego.
     */
    private String log;

    /**
     * Vector de jugadores que participan en el juego.
     */
    private ArrayList<Player> players;

    /**
     * Vector de monstruos presentes en el juego.
     */
    private ArrayList<Monster> monsters;

    /**
     * Laberinto en el que se desarrolla el juego.
     */
    private Labyrinth labyrinth;
    
    /**
     * Constructor de la clase Game, que inicializará los jugadores a jugar 
     * (definiendo el jugador que inicia la partida) y el estado del laberinto, 
     * que incluye su propia construcción, es decir, definir los bloques,
     * los monstruos y sus posiciones.
     * 
     * @param nplayers Numero de jugadores
     */
    public Game (int nplayers){
        // Definimos casilla de salida
        int exitRow=Dice.randomPos(ROWS);
        int exitCol=Dice.randomPos(COLS);

        // Inicializamos los vectores de jugadores y monstruos
        this.players=new ArrayList<>();
        this.monsters=new ArrayList<>();
        
        // Creamos los nplayers y los introducimos en el vector de players
        for(int i=0; i<nplayers; i++)
            this.players.add(new Player(Character.forDigit(i, 10),
                                    Dice.randomIntelligence(), Dice.randomStrength()));
        
                
        // Definimos el jugador que empezará, es decir, el currentPlayer
        this.currentPlayerIndex=Dice.whoStarts(nplayers);
        this.currentPlayer=this.players.get(this.currentPlayerIndex);
        
        // Inicializamos la instancia de laberinto
        this.labyrinth= new Labyrinth(ROWS, COLS, exitRow, exitCol);
        // Se configura con bloques y mosntruos el laberinto
        this.configureLabyrinth();

        // Se distribuyen los jugadores por el laberinto
        this.labyrinth.spreadPlayers(this.players);
        
        // Inicializamos log
        this.log="- Game just started.\n";
    }
    
    /**
     * Indica si algún jugador ha ganado la partida, es decir, si ha finalizado
     * el juego
     * @return True si se finalizo el juego, false en caso contrario
     */
    public boolean finished(){
        return this.labyrinth.haveWinner();
    }
    
    /**
     * Método que estudia un turno completo del juego. Gestiona el paso de un turno a otro.
     * @param preferredDirection  Dirección a la que se pretende mover el jugador
     * @return True si se finalizo el juego, false en caso contrario
     */
    public boolean nextStep (Directions preferredDirection){
        
        // Reseteamos el log
        log = "";

        if (currentPlayer.dead()){
            // Si el jugador está muerto, se ve si se resucita o no
            manageResurrection();
        }
        else{

            // Se obtiene la dirección real a la que se mueve el jugador.
            // Si no se puede mover, se indica en el log
            Directions direction = this.actualDirection(preferredDirection);
            if (direction != preferredDirection){
                logPlayerNoOrders();
            }

            // Se mueve al jugador y se obtiene el monstruo que haya en la casilla
            Monster monster = this.labyrinth.putPlayer(direction, currentPlayer);


            // Se estudia si hay un monstruo en la casilla a la que se ha movido el jugador
            if (monster == null){
                logNoMonster();
            }
            else{
                // Se lleva a cabo el combate
                GameCharacter winner = combat(monster);
                // Se gestiona la recompensa
                manageReward(winner);
            }// if (monster == null)
            
            
        }// if (currentPlayer.dead())


        boolean endGame = finished();
        if (!endGame)   // Si no ha finalizado el juego, se pasa al siguiente
            nextPlayer();

        return endGame;
    }
    
    /**
     * Crea una intancia de la clase GameState a partir de los miembros de 
     * la clase Game
     * 
     * @return Una intancia de GameState con los datos del estado actual del juego
     */
    public GameState getGameState(){
        // Creemos las cadenas de caracteres del estado de los
        // jugadores y de los monstruos
        String infoPlayers="";
        String infoMonsters="";
        
        // Formato jugadores
        for (int i=0; i<this.players.size(); i++){
            infoPlayers+="- " + this.players.get(i).toString()+"\n";
        }
        
        // Formato monstruos
        for (int i=0; i<this.monsters.size(); i++){
            infoMonsters+="- " + this.monsters.get(i).toString()+"\n";
        }
        
        // Estado general del juego
        GameState estadoGeneral = new GameState (this.labyrinth.toString(), infoPlayers,
                                infoMonsters, this.currentPlayerIndex, this.finished(), this.log);
        
        return estadoGeneral;
    }
    
    /**
     * Se crea el laberinto a partir de los monstruos y bloques definidos
     */
    private void configureLabyrinth(){
        // Inicializamos el vector de monstruos y los añadimos al laberinto
        for (int i=0; i<NUM_MONSTERS; i++){
            Monster monstruo=new Monster ("Monster "+i, Dice.randomIntelligence(), Dice.randomStrength());
            this.monsters.add(monstruo);
            // Destacar que la primera variable indica fila y la segunda columna
            this.labyrinth.addMonster(INIT_MONSTERS[i][0], INIT_MONSTERS[i][1], monstruo);
        }
        
        // Añadimos los bloques al laberinto
        for (int i=0; i<NUM_BLOCKS; i++){
            this.labyrinth.addBlock((Orientation)INIT_BLOCKS[i][0], (int)INIT_BLOCKS[i][1], 
                    (int)INIT_BLOCKS[i][2], (int)INIT_BLOCKS[i][3]);
        }
    }
    
    /**
     * Se indica que el turno pasa al siguiente jugador
     */
    private void nextPlayer(){
        // Pasamos al siguiente jugador según el índice del anterior
        this.currentPlayerIndex=(this.currentPlayerIndex+1) % this.players.size();
        this.currentPlayer=this.players.get(this.currentPlayerIndex);
    }
    
    /**
     * Método que mueve de hecho al jugador. Se intenta que se mueva a la casilla
     * indicada por la dirección, y si no es posible, se moverá en otra dirección.
     * 
     * @param preferredDirection  Dirección a la que se pretende mover el jugador
     * @return   Dirección a la que se moverá el jugador (si es válida)
     */
    private Directions actualDirection(Directions preferredDirection){
        
        // Se obtiene la fila y columna actual del jugador
        int currentRow = this.currentPlayer.getRow();
        int currentCol = this.currentPlayer.getCol();

        // Direcciones posibles desde la casilla actual
        ArrayList<Directions> validMoves = this.labyrinth.validMoves(currentRow, currentCol);

        // Se mueve al jugador
        return currentPlayer.move(preferredDirection, validMoves);
    }
    
    /**
     * Método encargado de llevar a cabo un combate entre el jugador actual y un monstruo.
     * 
     * @param monster Monstruo con el que se va a combatir.
     * @return GameCharacter: Jugador o Monstruo, tipo de rol del ganador del combate.
     */
    private GameCharacter combat(Monster monster){
        int rounds=0;   // Inicializamos el número de rounds a 0

        // Suponemos que el jugador ganará, y empieza este atacando.
        GameCharacter winner=GameCharacter.PLAYER;
        boolean lose = monster.defend(currentPlayer.attack());


        // Bucle que simula el combate entre el jugador y el monstruo, de forma alternada.

        while (!lose && rounds<MAX_ROUNDS){ // Si el monstruo no ha muerto y no se han superado los rounds

            rounds++;   // Incrementamos el número de rounds
            
            // Suponemos que el monstruo ganará, y continúa este atacando.
            winner = GameCharacter.MONSTER;
            lose = currentPlayer.defend(monster.attack());

            if (!lose){ // Si el jugador no ha muerto

                // Suponemos que el jugador ganará, y continúa este atacando.
                winner = GameCharacter.PLAYER;
                lose = monster.defend(currentPlayer.attack());
            }
        } // while (!lose && rounds<MAX_ROUNDS)

        this.logRounds(rounds, MAX_ROUNDS);
        return winner;        
    }
    
    /**
     * Método que gestiona las recompensas que se producen al finalizar un combate.
     * Se encarga de actualizar el log con la información de si ha ganado el jugador
     * o el monstruo.
     * 
     * @param winner Tipo de personaje que ha ganado el combate, puede ser un jugador o un monstruo.
     * @see GameCharacter
     */
    private void manageReward (GameCharacter winner){
        if (winner == GameCharacter.PLAYER){
            this.currentPlayer.receiveReward();
            this.logPlayerWon();
        }
        else{
            this.logMonsterWon();
        }
    }
    
    /**
     * Método que gestiona la resurrección de un jugador al finalizar un combate.
     * Destacar que en el caso de resucitar se deberá actualizar dicho jugador
     * por un fuzzyplayer, haciendo uso del método actualizarPos de la clase 
     * Labyrinth
     * 
     * @see Labyrinth#convertToFuzzy
     */
    private void manageResurrection (){
        if (Dice.resurrectPlayer()){
            // Se resucita al jugador 
            this.currentPlayer.resurrect();
            this.logResurrected();
            
            // Se cambia la posición por el fuzzyplayer
            FuzzyPlayer fuzzy= new FuzzyPlayer(this.currentPlayer);
            this.players.set(this.currentPlayerIndex, fuzzy);
            this.labyrinth.convertToFuzzy(fuzzy);
        }            
        else
            this.logPlayerSkipTurn();
    }     
    
    /**
     * Se actualiza el log con la información de que el jugador actual
     * ha ganado el combate
     */
    private void logPlayerWon(){
        this.log+= "- Player "+this.currentPlayerIndex+" won the fight.\n";
    }
    
    /**
     * Se actualiza el log con la información de que el mosntruo
     * ha ganado el combate al jugador actual
     */
    private void logMonsterWon(){
        this.log+= "- Monster won the fight.\n";
    }
    
    /**
     * Se actualiza el log con la información de que el jugador actual
     * ha ganado resucitado
     */
    private void logResurrected(){
        this.log+= "- Player "+this.currentPlayerIndex+" resurrected as fuzzy.\n";
    }
    
    /**
     * Se actualiza el log con la información de que el jugador actual
     * ha perdido su turno por estar muerto
     */
    private void logPlayerSkipTurn(){
        this.log+= "- Player "+this.currentPlayerIndex+" skipped turn (is dead).\n";
    }
    
    /**
     * Se actualiza el log con la información de que el jugador actual
     * ha intentado ejecutar alguna acción no permitida y no se ha producido ningún
     * cambio
     */
    private void logPlayerNoOrders(){
        this.log+= "- Player "+this.currentPlayerIndex+" didn't follow orders, it was not possible.\n";
    }
    
    /**
     * Se actualiza el log con la información de que el jugador actual
     * se ha desplazado a una casilla vacía o no ha sido posible el desplazamiento
     */
    private void logNoMonster(){
        this.log+= "- Player "+this.currentPlayerIndex+" moved to an empty square or it was not possible to move.\n";
    }
    
    /**
     * Se actualiza el log con la información de que se han producido rounds
     * de max rounds en el combate actual
     * 
     * @param rounds Número de rounds que se han producido
     * @param max Número máximo de rounds que puede durar un combate
     */
    private void logRounds(int rounds, int max){
        this.log+= "- Rounds: "+rounds+"/"+max+".\n";
    }
}
